src/phoenix/trace/openai/instrumentor.py

```diff
@@ -0,0 +1,210 @@
+import datetime
+import json
+from enum import Enum
+from inspect import signature
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterator, List, Optional, Tuple
+
+from phoenix.trace.schemas import (
+    Message,
+    SpanAttributes,
+    SpanEvent,
+    SpanException,
+    SpanKind,
+    SpanStatusCode,
+)
+from phoenix.trace.semantic_conventions import (
+    INPUT_MIME_TYPE,
+    INPUT_VALUE,
+    LLM_FUNCTION_CALL,
+    LLM_INPUT_MESSAGES,
+    LLM_INVOCATION_PARAMETERS,
+    LLM_TOKEN_COUNT_COMPLETION,
+    LLM_TOKEN_COUNT_PROMPT,
+    LLM_TOKEN_COUNT_TOTAL,
+    MESSAGE_CONTENT,
+    MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON,
+    MESSAGE_FUNCTION_CALL_NAME,
+    MESSAGE_NAME,
+    MESSAGE_ROLE,
+    OUTPUT_MIME_TYPE,
+    OUTPUT_VALUE,
+    MimeType,
+)
+from phoenix.trace.utils import get_stacktrace, import_package
+
+from ..tracer import Tracer
+
+if TYPE_CHECKING:
+    from openai.openai_response import OpenAIResponse
+
+
+class RequestType(Enum):
+    CHAT_COMPLETION = "chat_completion"
+    COMPLETION = "completion"
+    EMBEDDING = "embedding"
+
+
+INSTRUMENTED_ATTRIBUTE_NAME = "is_instrumented_with_openinference_tracer"
+
+
+class OpenAIInstrumentor:
+    def __init__(self, tracer: Optional[Tracer] = None) -> None:
+        """Instruments your OpenAI client to automatically create spans for each API call.
+
+        Args:
+            tracer (Optional[Tracer], optional): A tracer to record and handle spans. If not
+            provided, the default tracer will be used.
+        """
+        self._tracer = tracer or Tracer()
+
+    def instrument(self) -> None:
+        """
+        Instruments your OpenAI client.
+        """
+        openai = import_package("openai")
+        is_instrumented = getattr(
```

Comment:
`hasattr` also works

Reply:
addressed in [d845072](https://github.com/Arize-ai/phoenix/pull/1550/commits/d845072f5bbba851353f3ee1a6a1b3a6c55837bb)